File: examples/scripts/error_demo.py
#!/usr/bin/env python3
"""
Error demo skill implementation
Demonstrates colored error output using ANSI escape codes
"""
import sys
import json

# ANSI color codes
class Colors:
    RED = '\033[91m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    BLUE = '\033[94m'
    MAGENTA = '\033[95m'
    CYAN = '\033[96m'
    WHITE = '\033[97m'
    GRAY = '\033[90m'
    BOLD = '\033[1m'
    DIM = '\033[2m'
    RESET = '\033[0m'

def format_error(error_type, message, suggestions=None, context=None):
    """Format an error with colors and suggestions"""
    output = []
    
    # Error header with type indicator
    type_indicators = {
        'auth': 'AUTH',
        'network': 'NETWORK',
        'file': 'FILE', 
        'input': 'INPUT',
        'system': 'SYSTEM',
        'tool': 'TOOL'
    }
    
    indicator = type_indicators.get(error_type, 'ERROR')
    output.append(f"{Colors.RED}✗ {indicator}: {message}{Colors.RESET}")
    
    # Add context if provided
    if context:
        output.append(f"{Colors.GRAY}Context: {context}{Colors.RESET}")
    
    # Add suggestions if any
    if suggestions:
        output.append("")
        output.append(f"{Colors.BLUE}Suggestions:{Colors.RESET}")
        for suggestion in suggestions:
            output.append(f"  {Colors.GRAY}•{Colors.RESET} {Colors.WHITE}{suggestion}{Colors.RESET}")
    
    return "\n".join(output)

def get_error_examples():
    """Get all error type examples"""
    return {
        'auth': {
            'message': 'Authentication token has expired',
            'suggestions': [
                "Run 'q login' to authenticate",
                "Check your internet connection", 
                "Verify your AWS credentials"
            ]
        },
        'network': {
            'message': 'Failed to connect to API server',
            'context': 'Endpoint: https://api.example.com',
            'suggestions': [
                "Check your internet connection",
                "Try again in a few moments",
                "Verify the service is available"
            ]
        },
        'file': {
            'message': 'Permission denied',
            'context': 'Path: /etc/secure/config.json',
            'suggestions': [
                "Check if the file exists",
                "Verify file permissions",
                "Ensure the directory is accessible"
            ]
        },
        'input': {
            'message': 'Invalid command syntax: missing required argument',
            'context': 'Command: /example --missing-arg',
            'suggestions': [
                "Check the command syntax",
                "Use --help for usage information",
                "Verify all required parameters are provided"
            ]
        },
        'system': {
            'message': 'Internal system error occurred',
            'context': 'Component: session_manager',
            'suggestions': [
                "Restart the application",
                "Check system logs",
                "Contact support if the issue persists"
            ]
        },
        'tool': {
            'message': 'Tool execution timed out after 30 seconds',
            'context': 'Tool: git',
            'suggestions': [
                "Check tool permissions",
                "Verify tool dependencies are installed", 
                "Try running the tool manually"
            ]
        }
    }

def main():
    """Main execution function"""
    try:
        # Parse parameters from command line
        params = {}
        if len(sys.argv) > 1:
            params = json.loads(sys.argv[1])
        
        error_type = params.get('error_type')
        examples = get_error_examples()
        
        if error_type and error_type in examples:
            # Show specific error type
            example = examples[error_type]
            print(format_error(
                error_type,
                example['message'],
                example.get('suggestions'),
                example.get('context')
            ))
        else:
            # Show all error types
            print(f"{Colors.BOLD}{Colors.MAGENTA}Error Display Demo{Colors.RESET}")
            print()
            
            for err_type, example in examples.items():
                print(f"{Colors.BOLD}{Colors.MAGENTA}{err_type.upper()}:{Colors.RESET}")
                print(format_error(
                    err_type,
                    example['message'],
                    example.get('suggestions'),
                    example.get('context')
                ))
                print()
            
            print(f"{Colors.BLUE}Use error_type parameter to see specific examples{Colors.RESET}")
            print(f"{Colors.GRAY}Example: @error-demo auth{Colors.RESET}")
    
    except Exception as e:
        print(f"{Colors.RED}Error in error-demo skill: {e}{Colors.RESET}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()

File: crates/chat-cli/src/cli/chat/cli/error_demo.rs
use std::io::Write;

use clap::Args;
use eyre::Result;

use crate::cli::chat::ChatState;
use crate::cli::chat::ConversationState;
use crate::os::Os;
use crate::theme::{ErrorDisplay, ErrorType};

/// Demonstrate colored error output formatting
#[derive(Debug, Args)]
pub struct ErrorDemoArgs {
    /// Type of error to demonstrate
    #[arg(value_enum)]
    pub error_type: Option<DemoErrorType>,
}

#[derive(Debug, Clone, clap::ValueEnum)]
pub enum DemoErrorType {
    Auth,
    Network,
    File,
    Input,
    System,
    Tool,
}

impl ErrorDemoArgs {
    pub async fn execute(
        &self,
        _os: &mut Os,
        conversation_state: &mut ConversationState,
    ) -> Result<ChatState> {
        match self.error_type {
            Some(DemoErrorType::Auth) => {
                let error = ErrorDisplay::auth_error("Authentication token has expired");
                writeln!(conversation_state.stderr, "{}", error)?;
            },
            Some(DemoErrorType::Network) => {
                let error = ErrorDisplay::network_error("Failed to connect to API server")
                    .with_context("Endpoint: https://api.example.com");
                writeln!(conversation_state.stderr, "{}", error)?;
            },
            Some(DemoErrorType::File) => {
                let error = ErrorDisplay::file_error("Permission denied", Some("/etc/secure/config.json"));
                writeln!(conversation_state.stderr, "{}", error)?;
            },
            Some(DemoErrorType::Input) => {
                let error = ErrorDisplay::input_error("Invalid command syntax: missing required argument")
                    .with_context("Command: /example --missing-arg");
                writeln!(conversation_state.stderr, "{}", error)?;
            },
            Some(DemoErrorType::System) => {
                let error = ErrorDisplay::new(ErrorType::System, "Internal system error occurred")
                    .with_suggestion("Restart the application")
                    .with_suggestion("Check system logs")
                    .with_context("Component: session_manager");
                writeln!(conversation_state.stderr, "{}", error)?;
            },
            Some(DemoErrorType::Tool) => {
                let error = ErrorDisplay::tool_error("Tool execution timed out after 30 seconds", Some("git"));
                writeln!(conversation_state.stderr, "{}", error)?;
            },
            None => {
                // Show all error types
                use crate::theme::formatter;
                let fmt = formatter();
                
                writeln!(conversation_state.stdout, "{}", fmt.header("Error Display Demo"))?;
                writeln!(conversation_state.stdout)?;
                
                let error_types = [
                    ("auth", ErrorDisplay::auth_error("Sample authentication error")),
                    ("network", ErrorDisplay::network_error("Sample network error")),
                    ("file", ErrorDisplay::file_error("Sample file error", Some("/path/to/file"))),
                    ("input", ErrorDisplay::input_error("Sample input error")),
                    ("system", ErrorDisplay::new(ErrorType::System, "Sample system error")),
                    ("tool", ErrorDisplay::tool_error("Sample tool error", Some("example_tool"))),
                ];
                
                for (name, error) in error_types {
                    writeln!(conversation_state.stdout, "{}", fmt.emphasis(format!("{}:", name.to_uppercase())))?;
                    writeln!(conversation_state.stdout, "{}", error)?;
                }
                
                writeln!(conversation_state.stdout, "{}", fmt.info("Use --error-type <type> to see specific error examples"))?;
            },
        }
        
        Ok(ChatState::WaitingForInput)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_error_demo_args_creation() {
        let args = ErrorDemoArgs { error_type: None };
        assert!(args.error_type.is_none());
        
        let args = ErrorDemoArgs { error_type: Some(DemoErrorType::Auth) };
        assert!(matches!(args.error_type, Some(DemoErrorType::Auth)));
    }

    #[test]
    fn test_demo_error_types() {
        let types = [
            DemoErrorType::Auth,
            DemoErrorType::Network,
            DemoErrorType::File,
            DemoErrorType::Input,
            DemoErrorType::System,
            DemoErrorType::Tool,
        ];
        
        // Just verify all types can be created
        for error_type in types {
            let args = ErrorDemoArgs { error_type: Some(error_type) };
            assert!(args.error_type.is_some());
        }
    }
}
